### ISSUE TITLE
Inconcistency in behaviour with constrained Typevars and Union

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#2248]
**Describe the bug**
Pyright seems to have some unexpected behaviour related to union types and typevars with constraints, At least comparatively to mypy.

**To Reproduce**
Consider the code below: Pyright will output the following
```
  test.py:9:13 - info: Type of "a" is "Foo[int]"
  test.py:12:13 - info: Type of "b" is "Foo[str]"
  test.py:15:13 - info: Type of "c" is "Foo[int | str]"
  test.py:26:13 - info: Type of "bar" is "Foo[int]"
  test.py:34:13 - info: Type of "baz" is "Foo[str]"
  test.py:42:13 - info: Type of "qux" is "Foo[T@foo]
  ```

**Expected behavior**
There are a few things which stand out to me:

Firstly type of `c`, mypy rejects a union of constraints unless the union is expliclty in the constraints, is this intended behaviour?

Regardless or not, the type of `qux` is interesting, as it is left unbounded. given `c` is accepted should this not also be `Foo[int | str]`?

**Screenshots or Code**
```python
from typing import Any, Callable, Generic, TypeVar, Union

T = TypeVar('T', int, str)

class Foo(Generic[T]):
    ...

a = Foo[int]()
reveal_type(a)

b = Foo[str]()
reveal_type(b)

c = Foo[Union[int, str]]()
reveal_type(c)


def foo(func: Callable[[T], Any]) -> Foo[T]:
    ...


@foo
def bar(x: int) -> None:
    ...

reveal_type(bar)


@foo
def baz(x: str) -> None:
    ...


reveal_type(baz)


@foo
def qux(x: Union[int, str]) -> None:
    ...


reveal_type(qux)

```
![image](https://user-images.githubusercontent.com/8677174/131680371-c88a3a79-6795-48f7-b49d-b2348916bbf3.png)

**VS Code extension or command-line**
pyright 1.1.165

**Additional context**
MyPy (0.910) refuses to accept the union as valid.

```
test.py:9: note: Revealed type is "test.Foo[builtins.int*]"
test.py:12: note: Revealed type is "test.Foo[builtins.str*]"
test.py:14: error: Value of type variable "T" of "Foo" cannot be "Union[int, str]"
test.py:15: note: Revealed type is "test.Foo[Union[builtins.int, builtins.str]]"
test.py:26: note: Revealed type is "test.Foo[builtins.int*]"
test.py:34: note: Revealed type is "test.Foo[builtins.str*]"
test.py:37: error: Value of type variable "T" of "foo" cannot be "Union[int, str]"
test.py:42: note: Revealed type is "test.Foo[Union[builtins.int, builtins.str]]"
```
If I add `Union[int, str]` to the constraints of `T` mypy will no-longer complain however pyright behaves exactly the same as before.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.166, which I just published. It will also be included in the next release of pylance.
